### PR TITLE
[DRMAtomic] Setting DRM property by name instead of by hardcoded enum value

### PR DESCRIPTION
**Pull request by slattann**
This patch implements setting DRM property like Scaling_Filter by
validating actual supported values than by hard coding through local
enums.
 1) Added new API 'AddPropertyValueByName'
 2) removed the enum defitiion that hard codes the property values

Signed-off-by: Sameer Lattannavar <sameer.lattannavar@intel.com>

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:258]: This might be easier to read if you used `continue` when the name does not match. The same for the other `strcmp()` below.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:274]: You probably do not want to return `true` if the property or the enum value you are trying to set does not exist.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:58]: Btw. what will happen if DRM rejects these CRTC coordinates? That is, KMS atomic commit fails.

Not all planes might support values different from (0, 0, hdisplay, vdisplay).

**Review by ppaalanen — CHANGES_REQUESTED** [review]
A quick look on this patch in isolation, that is, not knowing anything about the surrounding code, gives me the impression that this is correct usage of the KMS UAPI.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (slattann) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp:58]: The idea here is to match the way GL is doing to make frame centralize after upscaling. I am not sure if there are some planes do not allow to request drawing from position other than (0, 0).

Comment (ppaalanen) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp:58]: Traditionally all primary planes refuse if you don't fill the whole plane exactly with the FB.

Any driver could refuse it for any reason, so there needs to be a fallback, but that applies to atomic KMS in general.

**Comment by ppaalanen** [comment]
What hardware and driver have you tested this on?

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (slattann) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp:58]: I got it. With that change user experience may not be good I thought. Upscaled frame will be drawn on left top corner and right bottom corner will be painted black.
Attached the difference in the user experience for your reference. 
image
I felt right one is better in viewing the game content so I have added that code. 
We can do one check if 
` AddProperty(object, "CRTC_X", (diff_w / 2));
  AddProperty(object, "CRTC_Y", (diff_h / 2));` 
fail then we can set to (0, 0).  What do you suggest ?

**Comment by slattann** [comment]
> What hardware and driver have you tested this on?

Tested it on TigerLake + i915 driver.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (slattann) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp:58]: ok. I will keep  CRTC (X, Y) to (0, 0) as it was before.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (lrusak) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:248]: `drm_object` struct doesn't exist anymore. You'll have to use the `CDRMObject` class.

**Comment by lrusak** [comment]
I think something like this would be helpful here (untested)


```c++

bool CDRMObject::GetPropertyValue(std::string name, std::string type, uint64_t& value) const
{
  auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(), [&name](auto& prop) {
    return StringUtils::EqualsNoCase(prop->name, name);
  });

  if (property == m_propsInfo.end())
    return false;

  if (drm_property_type_is(property->get(), DRM_MODE_PROP_ENUM) != 0)
  {
    for (int j = 0; j < property->get()->count_enums; j++)
    {
      if (std::strcmp(property->get()->enums[j].name, type.c_str()) == 0)
      {
        value = property->get()->enums[j].value;
        return true;
      }
    }
  }

  return false;
}
```

use with
```c++
uint64_t value;
if (m_gui_plane->GetPropertyValue("SCALING_FILTER", "Nearest Neighbor", value);
  AddProperty(m_gui_plane, "SCALING_FILTER", value);
```

or something similar

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (slattann) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: Ok got it. I will fix it .

**Comment by slattann** [comment]
> I think something like this would be helpful here (untested)
> 
> ```c++
> bool CDRMObject::GetPropertyValue(std::string name, std::string type, uint64_t& value) const
> {
>   auto property = std::find_if(m_propsInfo.begin(), m_propsInfo.end(), [&name](auto& prop) {
>     return StringUtils::EqualsNoCase(prop->name, name);
>   });
> 
>   if (property == m_propsInfo.end())
>     return false;
> 
>   auto prop = property->get();
> 
>   if (drm_property_type_is(prop, DRM_MODE_PROP_ENUM) != 0)
>   {
>     for (int j = 0; j < prop->count_enums; j++)
>     {
>       if (std::strcmp(prop->enums[j].name, type.c_str()) == 0)
>       {
>         value = prop->enums[j].value;
>         return true;
>       }
>     }
>   }
> 
>   return false;
> }
> ```
> 
> use with
> 
> ```c++
> uint64_t value;
> if (m_gui_plane->GetPropertyValue("SCALING_FILTER", "Nearest Neighbor", value))
>   AddProperty(m_gui_plane, "SCALING_FILTER", value);
> ```
> 
> or something similar
> 
> EDIT:
> are properties allowed to be 0? if not I would rather return the property value instead of bool and check `value != 0`

@lrusak, I have tested this for getting the right scaling_filter property value and it works. I have updated this PR. Kindly help to get this merged. Kernel Patches are pending to merge because of this blocker. Thanks

**Review by lrusak — APPROVED** [review]
LGTM

**Comment by lrusak** [comment]
jenkins build this please

**Review by garbear — APPROVED** [review]

**Comment by slattann** [comment]
Once again thank you so much to @ppaalanen  @lrusak  & @garbear  to help address everything to do right.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: No, that is not what I meant.

Setting the CRTC X, Y, W, H to center the integer-scaled image is fine!

However, you **must** prepare for the fact that the KMS driver may not be able to do that, so you need a *fallback* path.

**Comment by ppaalanen** [comment]
> ```c++
> StringUtils::EqualsNoCase
> ```

Does that mean case-insensitive match? DRM property names are case-sensitive. And yes, there is no common pattern for using upper-case letters in the names, so you do want to match the kernel spelling of a name literally.

0 is always an invalid id with KMS, so 0 cannot be a valid property id. Property values in general are allowed to be 0. It looks like 0 is a valid enum value:

> DRM_MODE_PROP_ENUM
>    Enumerated properties take a numerical value that ranges from 0 to the number of enumerated values defined by the property minus one, and associate a free-formed string name to each value. Applications can retrieve the list of defined value-name pairs and use the numerical value to get and set property instance values.

from https://www.kernel.org/doc/html/latest/gpu/drm-kms.html#property-types-and-blob-property-support

Otherwise that example code from @lrusak looks good for what I can understand of C++, though maybe `GetPropertyValue` might be confused with getting a property's current value from KMS rather than getting the uint64 corresponding to an enum value name?

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:49]: Does this even build? How can you say you tested this if it doesn't build?

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:57]: As I said, setting the CRTC X, Y, W, H to center the integer-scaled image is fine. You just need to prepare for the fact that the KMS driver may not support that, particularly if the KMS plane happens to be the primary plane.

If that does not work, the it is very likely that you cannot have `"CRTC_W", (mar_scale_factor * m_width)` or `"CRTC_H", (mar_scale_factor * m_height)` either. How do you handle it when the KMS driver refuses?

It looks like `CDRMAtomic::DrmAtomicCommit()` has literally no fallback plan. It attempts a test-only commit, yes, but if that fails, you are supposed to try something else until it succeeds. Right now it looks like the user would be left with a black or frozen screen.

So you would try with the centered integer-scaled image first. If that fails, then must try again without integer scaling.

**Review by ppaalanen — CHANGES_REQUESTED** [review]

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (slattann) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp:49]: My bad, some garbage code left before pushing into PR, I will fix it. Thanks for pointing it out.

**Review by emersion — APPROVED** [review]
This looks like correct kernel uAPI usage to me. Thanks for fixing this!

**Comment by lrusak** [comment]
jenkins build this please

**Comment by slattann** [comment]
Since Kernel patches have been merged into [drm-tip](https://github.com/freedesktop/drm-tip/commit/5c759eda9b04cd1047b4cda8ac1eaadf8a9e4fce), can we please merge this PR?

**Comment by howie-f** [comment]
jenkins build this please

**Comment by howie-f** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Comment by ppaalanen** [comment]
@slattann, since your comment sounds like you believe the kernel patch landed, let me ask you this.

Quoting https://www.kernel.org/doc/html/latest/gpu/drm-uapi.html#open-source-userspace-requirements :

> The kernel patch can only be merged after all the above requirements are met, but it must be merged to either drm-next or drm-misc-next before the userspace patches land.

Is drm-tip one of the drm-next or drm-misc-next, or is drm-tip just a throw-away integration-testing branch from which commits can disappear at any time?

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp
Comment (ppaalanen) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp:57]: Please, explain how this review comment has been addressed, or alternatively why it does not need addressing?

**Comment by emersion** [comment]
>Is drm-tip one of the drm-next or drm-misc-next, or is drm-tip just a throw-away integration-testing branch from which commits can disappear at any time?

drm-tip is a throw-away integration-testing branch. Each time you push to drm-next or drm-misc-next (or other repos), the drm-tip integration branch is re-generated. More info: https://drm.pages.freedesktop.org/maintainer-tools/drm-tip.html

**Comment by DaveTBlake** [comment]
jenkins build this please
@lrusak @garbear you both have approved, one of you like to hit merge (or not, and why) I know nothing about DRM etc.

**Comment by lrusak** [comment]
> jenkins build this please
> @lrusak @garbear you both have approved, one of you like to hit merge (or not, and why) I know nothing about DRM etc.

just waiting for a successful jenkins build :)

**Comment by DaveTBlake** [comment]
jenkins build this please

**Comment by garbear** [comment]
@ppaalanen @emersion @slattann Is now a good time to merge? We're wrapping up PRs into our last alpha. We gladly accept less-that-perfect solutions this early, because we'll have a long time to sort out any changes or fixes that need to be made.

**Comment by DaveTBlake** [comment]
@garbear it sounded like @lrusak was ready to mege this, so it would have been merged expects for jenkins being difficult. Once green I'll merge this.

**Comment by DaveTBlake** [comment]
Right build separately jenkins CI is happy. See https://jenkins.kodi.tv/view/Helpers/job/BuildMulti-PR-Manually/400/ so hitting that merge button.

**Pull request merged**
